WordPress/gutenberg
# Block API: Add role for 'internal' attributes not copied on duplication

Closes #29693
Alternative to #32604 

## Description


This PR:
* Adds handling for a new `internal` attribute experimentalRole. Attributes registered with this role will not be copied on block duplication.
* Refactors `__internalWidgetID` to use the new experimental role
 
Example fragment of block.json:

```
            "attributes": {
		"myInternalAttribute": {
			"type": "string",
			"default": "my-default-value",
			"__experimentalRole": "internal"
		}
	},
```

For just one example of how this might be used: the Jetpack `Pay with Paypal` block uses a `productId` attribute to identify the actual product record referenced by the block. This data is meant to be internal to the block and should not be exposed to the user. When the user duplicates the block, we want to copy the other attributes used to store price info and other details, but create a brand new product record/productId.

This is similar conceptually to #32604 with a few changes:
* The use of `internal` terminology rather than `duplicable` or `unique`, which implies additional constraints
* The use of the `__experimentalRole` and associated utilities (thanks @ntsekouras for the suggestion)

## Notes

* In other regards, an attribute with this role acts like a normal attribute
    * ~~It respects any configured default values~~ Default values are not filled in on block duplication either; support for this could be easily added if a viable use case becomes apparent, though
    * It can be managed as normal via `setAttributes`
    * It can apply to any type of attribute (ie it is not necessarily an id field)
* While you can give this role to an attribute with a "source"/"selector" configuration, it is not meaningful to do so. For example, adding it to the `content` attribute of the [Code block](https://github.com/WordPress/gutenberg/blob/a84c037a0e62a344005054102544c34d7b970a6b/packages/block-library/src/code/block.json#L9) would not make sense, because although the attribute would not be copied on duplication, it will populate with the same value from the html.
* Because this is experimental and also relies on experimental features, I have not added documentation; happy to do so if that is appropriate!
* A note about Reusable blocks:
Internal attributes are not removed when you convert a block into a Reusable block. I don’t know a ton about Reusable blocks, so calling this out so that someone more familiar can correct me; my understanding is that multiple instances of a Reusable block are not independent of each other (ie editing one copy affects them all), so internal attributes should also be shared.

## How has this been tested?



* Play around with adding `__experimentalRole: 'internal'` to any block attribute. 
* Insert the block into a new post, change its attributes, and duplicate it
* Verify that the internal attributes were not copied, but all other attributes were
* ~~Test with attributes that have a `default` value and ensure that the default is respected in the duplicated block~~
* Test Copy+Paste using both the `Copy` menu item, and by keyboard commands
* Test duplicating a block with internal attributes in the widget editor
* Test adding and editing widgets in both the Widget editor and the Customizer
  * Test moving widgets between widget areas/side panels
  * Test a server-side rendered block in the widget editor and make sure the preview is displayed correctly (Basically verify that #29197 is still fixed with the new implementation)

## Screenshots 

## Types of changes





## Checklist:
